cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs

```diff
@@ -0,0 +1,264 @@
+use cloudevents::event::SpecVersion;
+use cloudevents::message::{
+    BinaryDeserializer, BinarySerializer, Encoding, MessageAttributeValue, MessageDeserializer,
+    Result, StructuredDeserializer, StructuredSerializer,
+};
+use cloudevents::{message, Event};
+use rdkafka::message::{BorrowedMessage, Headers, Message, OwnedMessage};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::str;
+
+pub struct ConsumerRecordDeserializer {
+    pub(crate) headers: HashMap<String, Bytes>,
+    pub(crate) payload: Bytes,
+}
+
+impl ConsumerRecordDeserializer {
+    pub fn owned_new(message: OwnedMessage) -> ConsumerRecordDeserializer {
+        let mut resp_des = ConsumerRecordDeserializer {
+            headers: HashMap::new(),
+            payload: Bytes::new(),
+        };
+        let headers = message.headers().unwrap();
+        for i in 0..headers.count() {
+            let header = headers.get(i).unwrap();
+            resp_des
+                .headers
+                .insert(header.0.to_string(), Bytes::copy_from_slice(header.1));
+        }
+
+        match message.payload() {
+            Some(s) => resp_des.payload = Bytes::copy_from_slice(s),
+            None => resp_des.payload = resp_des.payload,
+        }
+
+        resp_des
+    }
+
+    pub fn borrowed_new(message: &BorrowedMessage) -> ConsumerRecordDeserializer {
+        let mut resp_des = ConsumerRecordDeserializer {
+            headers: HashMap::new(),
+            payload: Bytes::new(),
+        };
+        let headers = message.headers().unwrap();
+        for i in 0..headers.count() {
+            let header = headers.get(i).unwrap();
+            resp_des
+                .headers
+                .insert(header.0.to_string(), Bytes::copy_from_slice(header.1));
+        }
+
+        match message.payload() {
+            Some(s) => resp_des.payload = Bytes::copy_from_slice(s),
+            None => resp_des.payload = resp_des.payload,
+        }
+
+        resp_des
+    }
+}
+
+impl BinaryDeserializer for ConsumerRecordDeserializer {
+    fn deserialize_binary<R: Sized, V: BinarySerializer<R>>(self, mut visitor: V) -> Result<R> {
+        if self.encoding() != Encoding::BINARY {
+            return Err(message::Error::WrongEncoding {});
+        }
+
+        let spec_version = SpecVersion::try_from(header_value_to_str!(self
+            .headers
+            .get("ce_specversion")
+            .unwrap())?)?;
+
+        visitor = visitor.set_spec_version(spec_version.clone())?;
+
+        let attributes = spec_version.attribute_names();
+
+        for (hn, hv) in self
+            .headers
+            .iter()
+            .filter(|(hn, _)| "ce_specversion" != **hn && hn.starts_with("ce_"))
+        {
+            let name = &hn["ce_".len()..];
+
+            if attributes.contains(&name) {
+                visitor = visitor.set_attribute(
+                    name,
+                    MessageAttributeValue::String(String::from(header_value_to_str!(hv)?)),
```

Comment:
If you change from the comments above the header value from `Bytes` to `Vec<u8>`, then you don't need this dance here to convert to string. You just need `String::from_utf8` (which also avoids another unnecessary copy you're doing here)

Reply:
No I made a mistake here. I changed the function in headers from `str::from_utf8` to `String::from_utf8`. No more unnecessary copy :)

Reply:
No you don't, because `hv` should be already a `Vec<u8>` with the proposed change above